### PR TITLE
ES memory write race fix and performance improvement

### PR DESCRIPTION
**Pull request by paramite**
This patch is fixing #57 and:
- allows sending multiple bulk writes in parallel
- also avoid sending whole records via channel in elasticsearch app
- adds pure go (no cgo wrapper) amqp1 transport, which is more performant than with using official qpid C wrapper

### CODE REVIEW

File: plugins/application/elasticsearch/main.go
Comment (csibbitt) [inline plugins/application/elasticsearch/main.go:136]: As discussed, probably missing a `go func{ }` around this for loop

File: .github/workflows/integration.yml
Comment (csibbitt) [inline .github/workflows/integration.yml:71]: Did this commit slip in by mistake? I don't object to it, it just seems out of place.

**Comment by csibbitt** [comment]
Hey @paramite thanks for getting to the bottom of this. I see in a commit comment that you did indeed find problems with concurrent.Map. I'm going to re-read it now to see if I can spot it out of interest.

Just a reminder that force-push is generally a bad idea especially for github. It makes the flow of commits vs. comments illogical, some commits disappear which makes things even harder to follow, and my local branch becomes invalid so I need to delete it and re-pull to get back in sync. In short - please try to avoid force-push. It's okay to rewrite history in private, but once it's been published, you should seek to build upon history, not rewrite it.

File: plugins/application/elasticsearch/main.go
Comment (pleimer) [reply plugins/application/elasticsearch/main.go:56]: It looks like he is wrapping the whole write process in the lock rather than just the map. There must have been writes from other go routines happening to the buffer between the check if the index already existed (L92) and writing to the buffer later in the ReceiveEvent() function. This wouldn't necessarily be a problem with the concurrent map but rather how it was being used. That's just my thoughts looking at it, maybe Martin can explain further

File: plugins/application/elasticsearch/main.go
Comment (pleimer) [inline plugins/application/elasticsearch/main.go:139]: interesting do you see a significant performance improvement using a worker pool?

**Review by pleimer — APPROVED** [review]
Looks good to me. Granted, generally worker pools aren't necessary with goroutines since they are so inexpensive to start-up and the runtime manages the amount that are created.

**Comment by pleimer** [comment]
In fact you could just write to ES inside the ReceiveEvent() function since that is already using it's own goroutine from the bus. It's kinda unnecessary to send it to Run() via channel before posting to ES. This may give even better performance. Just my speculation, I'm still fine with this patch

File: plugins/application/elasticsearch/main.go
Comment (csibbitt) [reply plugins/application/elasticsearch/main.go:56]: Hmmm, I'm still not sure I understand how that caused crashes, but upon re-reading with your explanation in mind, I do think I've spotted where logs were being lost! Between L96 and L102 in the old code; if two routines perform this at the same time, one of their results will be overwritten. This probably accounts for the losses I was seeing at high rates.

**Comment by csibbitt** [comment]
> Looks good to me. Granted, generally worker pools aren't necessary with goroutines since they are so inexpensive to start-up and the runtime manages the amount that are created.

If I understand correctly, you may be suggesting that instead of starting 3 goroutines that each wait on channel and call `es.client.Index` in response, it may be just as safe (and likely more performant) to have just the single channel reader and simply call `es.client.Index` itself inside a goroutine. @paramite has mentioned that the bottleneck now seems to be emptying memory into ES itself (bufferSize can help, but only to a point I guess). Do you think we can just let the go runtime decide how many workers is optimal? I'd certainly be interested in trying that to see how it works out.

**Review by csibbitt — APPROVED** [review]
I'm okay with merging this code, but if Paul's suggestions can lead to a simpler code, fewer tunable parameters, and more performance, then I think they are worth trying before we consider this work complete.

**Comment by pleimer** [comment]
> > Looks good to me. Granted, generally worker pools aren't necessary with goroutines since they are so inexpensive to start-up and the runtime manages the amount that are created.
> 
> If I understand correctly, you may be suggesting that instead of starting 3 goroutines that each wait on channel and call `es.client.Index` in response, it may be just as safe (and likely more performant) to have just the single channel reader and simply call `es.client.Index` itself inside a goroutine. @paramite has mentioned that the bottleneck now seems to be emptying memory into ES itself (bufferSize can help, but only to a point I guess). Do you think we can just let the go runtime decide how many workers is optimal? I'd certainly be interested in trying that to see how it works out.

Basically, but I was suggesting re-using the goroutine already spawned by the "bus" when it calls ReceiveEvent() here: https://github.com/infrawatch/sg-core/blob/23ade06b40b9e1e3eb138b9b9b8aa34f9f54cc46/pkg/bus/bus.go#L34.  I was suggesting placing the post to ES in the ReceiveEvent() function, making it so you would never need to create a channel to send events anywhere at all

But yeah that all depends on the configuration of Elasticsearch and how many connections it can handle. It might overwhelm ES if there are too many goroutines posting to it at the same time.

File: plugins/application/elasticsearch/main.go
Comment (pleimer) [reply plugins/application/elasticsearch/main.go:56]: It was crashing here: https://github.com/infrawatch/sg-core/blob/23ade06b40b9e1e3eb138b9b9b8aa34f9f54cc46/plugins/application/elasticsearch/main.go#L96

saying that the type of data at that index in the buffer was not a []string{} value, but a nil. It could be that another goroutine overwrote the value at that index before this type assertion happened. I'm not sure exactly where that would happen, just speculating

**Comment by paramite** [comment]
Hmm, interesting idea Paul. Will play with it in my lab env and in case update this PR. But anyway. This PR at least fixes the panics Chris was experiencing.

**Comment by paramite** [comment]
That was massive performance improvement indeed Paul!

**Comment by csibbitt** [comment]
> That was massive performance improvement indeed Paul!

Sweet. I look forward to re-testing this!  Any sign of problems with too many connections to ES? Are you counting messages into rsyslog == messages into ES ?

**Comment by paramite** [comment]
Hmm, so I'm tailing two log files and adding 2M records to both and after only following gets to the ES:

$ curl -X GET "http://127.0.0.1:9200/sglogs-lenovo_p720_rdo_13.2021.07.15/_count"
{"count":1381952,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0}}

That's not a good number .. sigh

**Comment by paramite** [comment]
Initially I thought that log record loss is happening somewhere in the elasticsearch app and the ES itself, but the print plugin routed to a file reports same number as indexed documents in ES. We are probably loosing data in the sg-core internals somewhere.

**Comment by csibbitt** [comment]
@paramite @pleimer Is there any reason we should not merge this? I see that it doesn't entirely fix the lost logs problem, but AFAICT it's an improvement, right? I'd like to merge this if possible, but since it's a bit stale I don't want to hit the button without a final check with you guys.

**Comment by vyzigold** [comment]
LGTM, the lost logs might have been dropped by the sg-bridge. If @paramite didn't see any lost logs with the quick patch I made to the sg-bridge like two weeks ago, then we should be good with log loss on sg-core side.

**Comment by leifmadsen** [comment]
@paramite if you have no objections to this, I think the consensus is to merge this down.

**Comment by paramite** [comment]
I need to update this a bit and then we can merge, if reviews will be positive.

File: .github/workflows/tests.yml
Comment (leifmadsen) [inline .github/workflows/tests.yml:38]:
```suggestion
          # golangci-lint command line arguments. https://staticcheck.io/docs/checks#SA4008 (The variable in the loop condition never changes, are you incrementing the wrong variable?)
          args: --exclude=SA4008
```

**Comment by leifmadsen** [comment]
Please don't force push over branches. Multiple commits is ok. I often use the commit notes when doing GA releases to make sure I capture changes in the release notes.

**Comment by paramite** [comment]
Leif I'm keeping all relevant changes in relevant commits, but when I need to remove a commit, then it does not make sense to revert it. Instead I remove it and force push. Right now I'm blindly trying to make appstream rsyslog work in the GH action. If I would leave all my trial/error commits, then the PR will be a mess.

**Comment by paramite** [comment]
Yay! CI fixed, cleaned the mess and this is ready for review now.

**Comment by paramite** [comment]
And yes, with Jaromir's sg-bridge patch or without using sg-bridge at all the log loss is 0%

File: plugins/application/elasticsearch/main.go
Comment (paramite) [reply plugins/application/elasticsearch/main.go:56]: Well the problem was that using concurrent.Map's method the data at index are locked only in the method itself, so yes, sometimes it happened that one goroutine fetched the data via Get and before it could work with that data other gouroutine deleted the index in map (because the map was not locked any more). That being said, concurrent.Map was not much ready for concurrency and I did not realize that when I wanted to use it for that purpose.

File: plugins/application/elasticsearch/main.go
Comment (paramite) [reply plugins/application/elasticsearch/main.go:142]: Fixed.

**Comment by csibbitt** [comment]
> ...force push. Right now I'm blindly trying to make appstream rsyslog work in the GH action. If I would leave all my trial/error commits, then the PR will be a mess.

1) We merge via "Squash and merge" so the main branch will only include a single commit rolling up your PR. I don't think 14 commits + 7 force push messages is any less messy than 21 commits when it comes to the PR screen itself. In fact it would consume less screen space. Heck, others might even learn something by watching you hack against the CI.

2) The solution to hacking against the CI without polluting your PR is not to use force-push, but instead create a second PR branched from the real one, hack on it with impunity, and then bring over the commit that works. Force pushing is an inconvenience to  anyone other than the one doing it (for the reasons I mentioned in July), so it's plainly inconsiderate, especially after multiple people have asked it to stop and provided reasons. I vote we disable on all branches since it promotes dangerous bad habits and provides nothing in return.

File: plugins/application/elasticsearch/main.go
Comment (csibbitt) [inline plugins/application/elasticsearch/main.go:167]: I'm confused by the history here. On July 13th Paul suggested eliminating the worker pool by re-using the goroutine already spawned by the call ReceiveEvent(). On July 14th you said "That was massive performance improvement indeed Paul!"  but I still see the workers code here. Did this approach need to be adandoned?

**Review by atyronesmith — APPROVED** [review]
Reviewed the concurrent map access code.  Looks good.

**Pull request merged**
